### ISSUE TITLE
Add `assembly ("memory-safe")` to the inline assembly blocks

### ISSUE THREAD

**Issue by hrkrshnn** [Vectorized/solady#187]
See https://docs.soliditylang.org/en/v0.8.17/assembly.html#memory-safety for information on the tag

If `via-ir` is used, the compiler may lift variables from stack to memory. This is only correct if all assembly block uses memory correctly, i.e., only use memory from `[0, 0x40)` (scratch space) and allocating memory starting from the current free memory pointer.

**Comment by Vectorized** [comment]
From experiments, this has no improvement on performance, so it is mostly for `via-ir` to be able to avoid stack too deep.

Need to think through on this. 
Some snippets abuse `0x40` and `0x60` temporarily. 
Some snippets allocate memory after populating it.
Not certain about how will the compiler handle such cases.

**Comment by hrkrshnn** [comment]
> From experiments, this has no improvement on performance

Yes, this is true. If the original code doesn't need lifting to memory, then there wouldn't be any difference.

The is more important when the library is used in contracts, and they need this memory lifting. It would be blocked by having Solady as a dependency.

> Some snippets abuse 0x40 and 0x60 temporarily.

Strictly speaking, this is abusing memory and can potentially lead to undefined behaviour. But in practice, I think it's unlikely these will have an effect.

>  Some snippets allocate memory after populating it.

This should be fine. Assuming the allocation follows the usual rules.

Maybe to give you some idea of what's going on: usually, the free memory pointer points to `0x80`. If there are, say 3 variables that needs to be lifted into memory, instead of starting the free memory pointer at `0x100`, the compiler instead starts at `0x80 + 3 * 0x20`. The three words starting from `0x80` until `0xa0` are then statically used to store these lifted variables. If you happened to use these slots in your inline assembly block, then it can lead to undefined behaviour.

In short, even if you mark a small subset of assembly blocks as "memory-safe", it's still useful for contracts using those functions. @Vectorized
